lib/mockito.dart

```diff
@@ -48,6 +53,132 @@ class Mock {
+/// the constructor.
+class FakeInvocation extends Invocation {
+  final Symbol memberName;
+  final Map<Symbol, dynamic> namedArguments;
+  final List<dynamic> positionalArguments;
+  final bool isGetter;
+  final bool isMethod;
+  final bool isSetter;
+
+  factory FakeInvocation(Invocation invocation) {
+    if (_typedArgs.isEmpty && _typedNamedArgs.isEmpty) {
+      throw new StateError("FakeInvocation called when no typed calls have been saved.");
+    }
+
+    // Handle named arguments first, so that we can provide useful errors for
+    // the various bad states. If all is well with the named arguments, then we
+    // can process the positional arguments, and resort to more general errors
+    // if the state is still bad.
+    var namedArguments = _reconstituteNamedArgs(invocation);
+    var positionalArguments = _reconstitutePositionalArgs(invocation);
+
+    _typedArgs.clear();
+    _typedNamedArgs.clear();
+
+    return new FakeInvocation._(
+        invocation.memberName,
+        positionalArguments,
+        namedArguments,
+        invocation.isGetter,
+        invocation.isMethod,
+        invocation.isSetter);
+  }
+
+  static Map<Symbol,dynamic> _reconstituteNamedArgs(Invocation invocation) {
+    var namedArguments = <Symbol, dynamic>{};
+    var _typedNamedArgSymbols = _typedNamedArgs.keys.map((name) => new Symbol(name));
+    invocation.namedArguments.forEach((name, arg) {
+      if (arg == null) {
+        if (!_typedNamedArgSymbols.contains(name)) {
+          // Incorrect usage of [typed], something like:
+          // `when(obj.fn(a: typed(any)))`.
+          throw new ArgumentError(
+              'A typed argument was passed in as a named argument named "$name", '
+              'but did not a value for its name. Each typed argument that is '
+              'passed as a named argument needs to specify the `name` argument. '
+              'For example: `when(obj.fn(x: typed(any, name: "x")))`.');
+        }
+      } else {
+        // Add each real named argument that was _not_ passed with [typed].
+        namedArguments[name] = arg;
+      }
+    });
+
+    _typedNamedArgs.forEach((name, arg) {
+      Symbol nameSymbol = new Symbol(name);
+      if (!invocation.namedArguments.containsKey(nameSymbol)) {
+        // Incorrect usage of [name], something like:
+        // `when(obj.fn(typed(any, name: 'a')))`.
+        throw new ArgumentError(
+            'A typed argument was declared with name $name, but was not passed '
+            'as an argument named $name.');
+      }
+      if (invocation.namedArguments[nameSymbol] != null) {
+        // Incorrect usage of [name], something like:
+        // `when(obj.fn(a: typed(any, name: 'b'), b: "string"))`.
+        throw new ArgumentError(
+            'A typed argument was declared with name $name, but a different '
+            'value (${invocation.namedArguments[nameSymbol]}) was passed as '
+            '$name.');
```

Comment:
I think the examples you give in the comments are super helpful. Can we try to incorporate them into the error. I don't think long errors are a bad thing when they help developers.